spring-rabbit/src/main/java/org/springframework/amqp/rabbit/annotation/MultiRabbitListenerAnnotationBeanPostProcessor.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2002-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.amqp.rabbit.annotation;
+
+import java.lang.reflect.Method;
+import java.util.Collection;
+
+import org.springframework.amqp.core.AbstractDeclarable;
+import org.springframework.amqp.core.Declarable;
+import org.springframework.amqp.rabbit.core.RabbitAdmin;
+import org.springframework.beans.factory.BeanFactory;
+import org.springframework.lang.NonNull;
+import org.springframework.util.StringUtils;
+
+/**
+ * An extension of {@link RabbitListenerAnnotationBeanPostProcessor} that associates the
+ * proper RabbitAdmin to the beans of Exchanges, Queues, and Bindings after they are
+ * created.
+ * <p>
+ * This processing restricts the {@link RabbitAdmin} according to the related
+ * configuration, preventing the server from automatic binding non-related structures.
+ *
+ * @author Wander Costa
+ */
+public class MultiRabbitListenerAnnotationBeanPostProcessor extends RabbitListenerAnnotationBeanPostProcessor {
+
+	public static final String CONNECTION_FACTORY_BEAN_NAME = "multiRabbitConnectionFactory";
+
+	public static final String CONNECTION_FACTORY_CREATOR_BEAN_NAME = "rabbitConnectionFactoryCreator";
+
+	private static final String DEFAULT_RABBIT_ADMIN_BEAN_NAME = "defaultRabbitAdmin";
+
+	private static final String RABBIT_ADMIN_SUFFIX = "-admin";
+
+	private BeanFactory beanFactory;
+
+	@Override
+	public void setBeanFactory(@NonNull BeanFactory beanFactory) {
+		this.beanFactory = beanFactory;
+		super.setBeanFactory(beanFactory);
+	}
+
+	@Override
+	protected Collection<Declarable> processAmqpListener(RabbitListener rabbitListener, Method method,
+			Object bean, String beanName) {
+		final Collection<Declarable> declarables = super.processAmqpListener(rabbitListener, method, bean, beanName);
+		final RabbitAdmin rabbitAdmin = resolveRabbitAdminBean(rabbitListener);
+		for (final Declarable declarable : declarables) {
+			if (declarable.getDeclaringAdmins().isEmpty() && declarable instanceof AbstractDeclarable) {
```

Comment:
I think we can add `setAdminsThatShouldDeclare` as a `default` method on `Declarable` and avoid the cast.

Reply:
I changed according to the suggestion.
Also, I suppose you also mean removing the condition `if (... && declarable instanceof AbstractDeclarable) {...}` to rely directly on the interface.

If that's the case, having a `default` method on `Declarable` might introduce space for a behavioral side-effect in MultiRabbit (declarables being processed by all admins, which I guess is not the main use-case). At `spring-amqp`, I see no impact, since all concrete classes eventually also extend `AbstractDeclarable`.

My thinking is that since these are [breaking changes](https://github.com/spring-projects/spring-amqp/pull/1111#pullrequestreview-346107472), why not introduce  `setAdminsThatShouldDeclare()` as a regular signature (non-`default`) and force concrete classes to implement it?
`Declarable`'s concretes must also implement `getDeclaringAdmins()` which makes me understand it's in the scope of its responsibilities.

What do you think about it?

Reply:
Yes; that would be ok (non-default) - I doubt users implement their own `Declarable` s anyway.